velox/vector/arrow/Bridge.cpp

```diff
@@ -315,68 +227,328 @@ const char* exportArrowFormatStr(const TypePtr& type) {
-    arrowArray.null_count = vector->getNullCount().value_or(-1);
+  // Whether filtering or reorder should be applied to the original elements.
+  bool changed() const {
+    return changed_;
+  }
+
+  template <typename F>
+  void apply(F&& f) const {
+    if (changed_) {
+      for (auto [offset, size] : ranges_) {
+        for (vector_size_t i = 0; i < size; ++i) {
+          f(offset + i);
+        }
+      }
+    } else {
+      for (vector_size_t i = 0; i < total_; ++i) {
+        f(i);
+      }
+    }
+  }
+
+  vector_size_t count() const {
+    if (!changed_) {
+      return total_;
+    }
+    vector_size_t ans = 0;
+    for (auto [_, size] : ranges_) {
+      ans += size;
+    }
+    return ans;
+  }
+
+  void clearAll() {
+    ranges_.clear();
+    changed_ = true;
+  }
+
+  void addRange(vector_size_t offset, vector_size_t size) {
+    ranges_.emplace_back(offset, size);
+  }
+
+ private:
+  std::vector<std::pair<vector_size_t, vector_size_t>> ranges_;
+  vector_size_t total_;
+  bool changed_ = false;
+};
+
+void gatherFromBuffer(
+    const Type& type,
+    const Buffer& buf,
+    const Selection& rows,
+    Buffer& out) {
+  auto src = buf.as<uint8_t>();
+  auto dst = out.asMutable<uint8_t>();
+  vector_size_t j = 0;
+  if (type.kind() == TypeKind::BOOLEAN) {
+    rows.apply([&](vector_size_t i) {
+      bits::setBit(dst, j++, bits::isBitSet(src, i));
+    });
   } else {
-    // If no nulls buffer, it means we have exactly zero nulls.
-    arrowArray.null_count = 0;
+    auto typeSize = type.cppSizeInBytes();
+    rows.apply([&](vector_size_t i) {
+      memcpy(dst + (j++) * typeSize, src + i * typeSize, typeSize);
+    });
   }
+}
 
-  switch (vector->encoding()) {
-    case VectorEncoding::Simple::FLAT:
-      exportFlatVector(vector, arrowArray, *bridgeHolder, pool);
+void exportNulls(
+    const BaseVector& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  if (!vec.nulls()) {
+    out.null_count = 0;
+    return;
+  }
+  if (!rows.changed()) {
+    holder.setBuffer(0, vec.nulls());
+    out.null_count = vec.getNullCount().value_or(-1);
+    return;
+  }
+  auto nulls = AlignedBuffer::allocate<bool>(out.length, pool);
+  gatherFromBuffer(*BOOLEAN(), *vec.nulls(), rows, *nulls);
+  holder.setBuffer(0, nulls);
+  out.null_count = -1;
+}
+
+void exportValues(
+    const BaseVector& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  out.n_buffers = 2;
+  if (!rows.changed()) {
+    holder.setBuffer(1, vec.values());
+    return;
+  }
+  auto values = vec.typeKind() == TypeKind::BOOLEAN
+      ? AlignedBuffer::allocate<bool>(out.length, pool)
+      : AlignedBuffer::allocate<uint8_t>(
+            out.length * vec.type()->cppSizeInBytes(), pool);
+  gatherFromBuffer(*vec.type(), *vec.values(), rows, *values);
+  holder.setBuffer(1, values);
+}
+
+void exportStrings(
```

Comment:
Github reviews are terrible =/ can't say if this function was just moved or if anything was modified in it...

Reply:
It's a quite dramatic change due to `Selection` so it's understandable that git hub did not detect it.  The code changes but I keep all the logic flow untouched though.